### ISSUE TITLE
Default trait for addrinfo?

### ISSUE THREAD

**Issue by saghul** [rust-lang-nursery/libc#158]
Hi!

Getting my feet wet with Rust here :-) I ran into a slight inconvenience when using a `addrinfo` struct for passing it as `hints` for `getaddrinfo`: different platforms have different fields, so fully initializing the struct would requite some `[cfg]` trickery, if I'm not mistaken.

So, I was thinking that adding the `Default` trait to `addrinfo` would help here, since I could write something like:

~~~~
let mut hints: libc::addrinfo = Default::default();
hints.ai_family = af;
hints.ai_socktype = typ;
hints.ai_protocol = proto;
~~~~

vs

~~~~
    let hints = libc::addrinfo {
        ai_family : af,
        ai_socktype : typ,
        ai_protocol : proto,
        ai_flags : 0,
        ai_addrlen : 0,
        ai_addr : ptr::null_mut(),
        ai_canonname : ptr::null_mut(),
        ai_next : ptr::null_mut(),
    };
~~~~

Does this make any sense? If so, I'd be happy to try and make a PR :-)

Thanks!

**Comment by alexcrichton** [comment]
Hello, and thanks for the report! Currently the design of this library is to avoid extra traits like `Default`, `PartialEq`, `Ord`, etc, for the types defined here. One reason is for codegen times (e.g. you probably don't need `Default` on *all* the types) and another is that this is otherwise just intended to be a "zero cost abstraction" library over libc in terms of runtime and compile-time (e.g. a giant header file).

The good news, however, is that you can do initialization as you would in C most of the time. For example:

```rust
let mut hints: libc::addrinfo = mem::zeroed();
hints.ai_family = af;
// ...

// or, "more unsafely"
let mut hints: libc::addrinfo = mem::uninitialized();
hints.ai_family = af;
// ...
```

Basically the idea is that you can create a generically zeroed version of the struct via `mem::zeroed` (which is similar to declaration + memset in C) or use `mem::uninitialized` (declaration with no memset). It's still a little unsafe, but you're guaranteed that there's no owned types in libc as all structs are `Copy` so in practice it should normally end up being "no more unsafe" than the FFI that's about to happen.

Hope that helps!

**Comment by saghul** [comment]
Hi Alex! I see now, thanks a lot for taking the time to explain :-)
